Fabric.Authorization.API/scripts/AzureAD-Migration.ps1

```diff
@@ -0,0 +1,252 @@
+                Write-DosMessage -Level Error -Message "Error authenticating user with Azure AD tenant $($tenantId). Please try again."
+                $authenticationFailed = $true
+            }
+            # Connect-AzureADTenant will not throw an exception if a valid username (i.e., user@domain, even if user is not in tenant
+            # and/or invalid password is entered). This catch block handles the 403 that is returned when Get-AzureADGroup fails
+            # for this scenario
+            catch [Microsoft.Open.AzureAD16.Client.ApiException] {
+                if ($_.Exception.ErrorCode -eq 403) {
+                    Write-DosMessage -Level Information -Message "403 error when retrieving Azure AD group."
+                    # reset cached credentials for this tenantId
+                    $credentials.Remove($tenantId)
+                    Write-DosMessage -Level "Information" -Message "Removed cached credentials for Azure AD Tenant $($tenantId)..."
+                    $authenticationFailed = $true
+                }
+                else {
+                    Write-DosMessage -Level Information -Message "Unexpected error when retrieving Azure AD group. $($_.Exception)"
+                    throw
+                }
+            }
+            catch {
+                $errorMsg = "Unexpected error while connecting to Azure AD: $($_.Exception)"
+                Write-DosMessage -Level Error -Message $errorMsg
+                throw $errorMsg
+            }
+        } while ($authenticationFailed -eq $true)
+    }
+
+    # disconnect from Azure AD
+    try {
+        Disconnect-AzureAD
+    }
+    catch {
+        $errorMsg = "There was an unexpected error error while disconnecting from Azure AD tenant: $($tenantId)"
+        Write-DosMessage -Level Error -Message $errorMsg
+        throw $errorMsg
+    }
+
+    if ($azureADGroups.Count -eq 1) {
+        return $azureADGroups[0]
+    }
+    elseif ($azureADGroups.Count -eq 0) {
+        $errorMsg = "No match found for SID $($groupSID) in Azure AD Tenant $($tenantId)."
+        Write-DosMessage -Level Information -Message $errorMsg
+        throw $errorMsg
+    }
+    else {
+        $errorMsg = "Multiple matches found for group SID $($groupSID)."
+        Write-DosMessage -Level Error -Message $errorMsg
+        throw $errorMsg
+    }
+}
+
+function Move-ActiveDirectoryGroupsToAzureAD {
+     param(
+        [Parameter(Mandatory=$true)]
+        [string] $connString
+    )
+
+    Write-DosMessage -Level Information -Message "Migrating AD groups to Azure AD..."
+
+    $allowedTenantIds = @()
+
+    try {
+        $allowedTenantIds = Get-AzureADTenants -installConfigPath $installConfigPath
+    }
+    catch {
+        Write-DosMessage -Level Fatal -Message "Error retrieving Azure AD Tenants. Check to ensure the installConfigPath ($($installConfigPath)) is correct."
+        throw
```

Comment:
this throw and the one below should be unnecessary if I understand how Write-DosMessage works.